internal/tool/command.go

```diff
@@ -0,0 +1,160 @@
+	"bytes"
+	"encoding/json"
+	"errors"
+	"os"
+	"os/exec"
+	"strings"
+
+	codacy "github.com/codacy/codacy-engine-golang-seed/v6"
+	docgen "github.com/codacy/codacy-semgrep/internal/docgen"
+	"github.com/samber/lo"
+)
+
+type SemgrepOutput struct {
+	Results []SemgrepResult `json:"results"`
+	Errors  []SemgrepError  `json:"errors"`
+}
+
+type SemgrepResult struct {
+	CheckID       string          `json:"check_id"`
+	Path          string          `json:"path"`
+	StartLocation SemgrepLocation `json:"start"`
+	EndLocation   SemgrepLocation `json:"end"`
+	Extra         SemgrepExtra    `json:"extra"`
+}
+
+type SemgrepLocation struct {
+	Line int `json:"line"`
+}
+
+type SemgrepExtra struct {
+	Message     string `json:"message"`
+	RenderedFix string `json:"rendered_fix,omitempty"`
+}
+
+type SemgrepError struct {
+	Message  string               `json:"message"`
+	Location SemgrepErrorLocation `json:"location"`
+}
+
+type SemgrepErrorLocation struct {
+	Path string `json:"path"`
+}
+
+func executeCommandForFiles(configFile *os.File, toolExecution codacy.ToolExecution, patternDescriptions *[]codacy.PatternDescription, language string, files []string) ([]codacy.Result, error) {
+	semgrepCmd := createCommand(configFile, toolExecution.SourceDir, language, files)
+
+	semgrepOutput, semgrepError, err := runCommand(semgrepCmd)
+	if err != nil {
+		return nil, errors.New("Error running semgrep: " + semgrepError + "\n" + err.Error())
+	}
+
+	output, err := parseCommandOutput(toolExecution.ToolDefinition, patternDescriptions, semgrepOutput)
+	if err != nil {
+		return nil, err
+	}
+	return output, nil
+}
+
+func createCommand(configFile *os.File, sourceDir, language string, files []string) *exec.Cmd {
+	params := createCommandParameters(language, configFile, files)
+	cmd := exec.Command("semgrep", params...)
+	cmd.Dir = sourceDir
+
+	return cmd
+}
+
+func createCommandParameters(language string, configFile *os.File, filesToAnalyse []string) []string {
+	// adding -json parameters
+	cmdParams := []string{
+		"-json", "-json_nodots",
+	}
+	// adding -lang parameter
+	cmdParams = append(
+		cmdParams,
+		"-lang", language,
+	)
+	// adding -rules parameter
+	cmdParams = append(
+		cmdParams,
+		"-rules", configFile.Name(),
+	)
+	// adding files to analyse
+	cmdParams = append(
+		cmdParams,
+		filesToAnalyse...,
+	)
+	return cmdParams
+}
+
+func runCommand(cmd *exec.Cmd) (string, string, error) {
+	var stderr bytes.Buffer
+	cmd.Stderr = &stderr
+	cmdOutput, err := cmd.Output()
+	if err != nil {
+		return "", stderr.String(), err
+	}
+	return string(cmdOutput), "", nil
+}
+
+func parseCommandOutput(toolDefinition codacy.ToolDefinition, patternDescriptions *[]codacy.PatternDescription, commandOutput string) ([]codacy.Result, error) {
+	var result []codacy.Result
+	scanner := bufio.NewScanner(strings.NewReader(commandOutput))
+	for scanner.Scan() {
+		output := scanner.Text()
+		result = appendToResult(result, patternDescriptions, output)
+	}
+
+	return result, nil
```

Comment:
I don't get what's the point of doing this. `commandOutput` is already a string, so can't you just parse it like you do in `appendToResult`?

Reply:
I think you are referring to using scan() vs just parsing the result as a whole. Here we are analysing the output of semgrep which will be a collection of both results and errors, and using the scan we are fetching each item sequentially and not interpreting it as a whole. I think that both ways are valid, its just a matter of preference, but correct me if im wrong.

Reply:
So if I understand correctly, the way we are doing this right now is actually not the best option since we are interpreting it as a JSON object per line (and it wouldn't be a valid JSON). So we shoulD unmarshal it directly into a domain struct like we do in `appendToResult`. 
If so, I think the only current "blocker" with that would be that we should interpret the output (which can be a lot), reproduce its structure into a struct in our code and then we would only get the results and errors out of it, because that's what's needed.

Reply:
If it's working right now, then my understanding is probably incorrect, so just keep it like that.